// addons-scanner-utils/dist/functions/download.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.downloadFileUpload = exports.downloadFile = exports.DownloadFileError = exports.DEFAULT_DOWNLOAD_FILENAME = void 0;
const node_os_1 = __importDefault(require("node:os"));
const node_path_1 = __importDefault(require("node:path"));
const node_fs_1 = require("node:fs");
const node_stream_1 = require("node:stream");
const promises_1 = require("node:stream/promises");
const api_1 = require("./api");
const auth_1 = require("./auth");
exports.DEFAULT_DOWNLOAD_FILENAME = 'input.xpi';
/**
 * Error thrown when file download operations fail.
 *
 * @property status - HTTP status code
 * @property extraInfo - Additional error details
 */
class DownloadFileError extends Error {
    constructor(message, extraInfo, status = 500) {
        super(message);
        this.name = 'DownloadFileError';
        this.status = status;
        this.extraInfo = extraInfo;
    }
    /**
     * Convert this error to ApiError format.
     */
    toApiError() {
        return (0, api_1.createApiError)({
            message: this.message,
            extraInfo: this.extraInfo,
            status: this.status,
        });
    }
}
exports.DownloadFileError = DownloadFileError;
const fetchToFile = (downloadURL, tmpDir, filename, headers) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        const response = yield fetch(downloadURL, { headers });
        if (!response.ok || !response.body) {
            throw new DownloadFileError(`unexpected response: ${response.statusText}`);
        }
        const filepath = node_path_1.default.join(tmpDir, filename);
        yield (0, promises_1.pipeline)(node_stream_1.Readable.fromWeb(response.body), (0, node_fs_1.createWriteStream)(filepath));
        return filepath;
    }
    catch (err) {
        if (err instanceof DownloadFileError) {
            throw err;
        }
        const message = err instanceof Error ? err.message : String(err);
        throw new DownloadFileError('failed to download file', message);
    }
});
/**
 * Download a file from the AMO API, authenticating with a JWT token computed
 * from the `AMO_JWT_ISS_KEY` and `AMO_JWT_SECRET` environment variables.
 *
 * @returns Promise resolving to the file path of the downloaded file
 * @throws DownloadFileError if the download fails
 */
const downloadFile = (_a) => __awaiter(void 0, [_a], void 0, function* ({ downloadURL, tmpDir = node_os_1.default.tmpdir(), filename = exports.DEFAULT_DOWNLOAD_FILENAME, env, }) {
    const token = (0, auth_1.makeJWT)({ env });
    return fetchToFile(downloadURL, tmpDir, filename, {
        Authorization: `JWT ${token}`,
    });
});
exports.downloadFile = downloadFile;
/**
 * Download a file from a URL to a temporary location.
 *
 * @returns Promise resolving to the file path of the downloaded file
 * @throws DownloadFileError if the download fails or URL origin doesn't match
 */
const downloadFileUpload = (_a) => __awaiter(void 0, [_a], void 0, function* ({ downloadURL, allowedOrigin, tmpDir = node_os_1.default.tmpdir(), filename = exports.DEFAULT_DOWNLOAD_FILENAME, }) {
    if (!downloadURL.startsWith(`${allowedOrigin}/`)) {
        throw new DownloadFileError(`download URL does not match allowed origin: ${allowedOrigin}`, undefined, 400);
    }
    return fetchToFile(downloadURL, tmpDir, filename);
});
exports.downloadFileUpload = downloadFileUpload;
